fix(amf): Handling PDU session establishment reject Scenarios

sreedharkumartn opened the pull request:
Signed-off-by: sreedharkumartn <sreedhar.kumar@wavelabs.ai>

<!--
    Tag your PR title with the components that it touches along with
    the type of change
    E.g. "fix(orc8r): Fix reindexer race condition" or "feat(agw) ..."
-->

## Summary
- Handled below Failure Cases:
-------------------------------------------------------------------------------------------------------------------------
Cause 27 – Missing or unknown DNN
This 5GSM cause is used by the network to indicate that the requested service was rejected by the external DN
because the DNN was not included although required or if the DNN could not be resolved.

Cause 28 – Unknown PDU session type
This 5GSM cause is used by the network to indicate that the requested service was rejected by the external DN
because the requested PDU session type could not be recognized or is not allowed.

Cause 43 - Invalid PDU session identity
Use case: If AMF receives new PDU Session Establishment Request and
                AMF have already another active PDU Session in progress with Same PDU Session Identity,
                AMF will not process PDU Session Establishment requests for 5 times, after that AMF rejects with reason 43.

-----------------------------------------------------------------------------------------------------------------------------------
- DNN optional IE support in PDU Session Establishment Request.

<!-- Enumerate changes you made and why you made them -->

## Test Plan
- Verified below Failures cases Using UERANSIM.
- Attached Pcap snap and log file of all failure cases.
1. PDU Session Establishment Failure with reason 27

Log file:
[mme_missing_dnn.log](https://github.com/magma/magma/files/7389235/mme_missing_dnn.log)

Pcap snap:
![missing_dnn](https://user-images.githubusercontent.com/89978170/138283804-d529b8f6-1025-49c7-a64d-3f84ea25db74.JPG)

2. PDU Session Establishment Failure with reason 28

Log file:
[mme_unknown_session_type.log](https://github.com/magma/magma/files/7389239/mme_unknown_session_type.log)

Pcap snap:
![unknown_pdu_session_type](https://user-images.githubusercontent.com/89978170/138283962-1790a1f2-0f2a-4d74-b8f9-972f4905aa1b.JPG)

3. PDU Session Establishment Failure with reason 43

Log file:
[mme_invalid_pdu_session_identity.log](https://github.com/magma/magma/files/7389242/mme_invalid_pdu_session_identity.log)

Pcap snap:
![invalid_pdu_session_identity](https://user-images.githubusercontent.com/89978170/138284093-cee0a1de-cee0-4b13-8561-3901b7b90b7f.JPG)

- Added UT cases to validate Failure Structure and validated.
![UT Cases](https://user-images.githubusercontent.com/89978170/138284294-ef00c73f-8007-4c4e-92e1-3216a46a8028.JPG)

<!--
    How did you test your change? How do you know it works?
    Add supporting screenshots, terminal pastes, etc. as necessary
-->

## Additional Information

- [ ] This change is backwards-breaking

<!--
    If this is a backwards-breaking change, document the upgrade instructions.
    All upgrade instructions for backwards-breaking changes will be aggregated
    in the next release's changelog so this is very important to fill out.
-->

sreedharkumartn committed:
fix(amf): Handling PDU session establishment reject Scenerios

Signed-off-by: sreedharkumartn <sreedhar.kumar@wavelabs.ai>

panyogesh commented on lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h:
```diff
@@ -268,6 +268,8 @@ typedef struct smf_context_s {
   protocol_configuration_options_t pco;
 
+  uint32_t count_duplicate_pdu_session_est_req;
+
```
Can you please rename it as pdu_session_est_req_count..(or duplicate_pdu_session_est_req_count)

sreedharkumartn replied:
Addressed review comment

panyogesh commented on lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp:
```diff
@@ -35,16 +35,23 @@ extern "C" {
+const uint32_t MAX_UE_INITIAL_PDU_SESSION_ESTABLISHMENT_REQ_ALLOWED = 5;
 
 int amf_max_pdu_session_reject(amf_ue_ngap_id_t ue_id, ULNASTransportMsg* msg);
+int amf_pdu_session_establishment_reject(
```
Can we have these definitions in another header files like :  "include/amf_smf_packet_handler.h"

sreedharkumartn replied:
created new header file as specified.

panyogesh commented on lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp:
```diff
@@ -374,14 +381,39 @@ int amf_smf_send(
+      M5GPduSessionType::ETHERNET == session_type) {
+    cause = M5GSmCause::UNKNOWN_PDU_SESSION_TYPE;
+  }
+
```
This function is getting bigger. Can you please split it like                                        _amf_smf_validate_params and then rest of regular functionality

sreedharkumartn replied:
Two new functions are created from this function. 1. For 5GSM Causes - amf_smf_validate_ie_params() and 2. For 5GMM Cause - amf_smf_validate_context().

panyogesh commented on lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp:
```diff
@@ -824,4 +885,182 @@ int amf_max_pdu_session_reject(
+ **                                                                        **
+ ***************************************************************************/
+
+int construct_pdu_session_reject_dl_req(
```
Mostly it seems will be used here. Can you please make it static. Also in test program call the parent function. Check the possibility.

sreedharkumartn replied:
This function is calling by two different files (amf_smf_send.cpp, test_amf_encode_decode.cpp). Right now we can't make it as static.

panyogesh commented on lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp:
```diff
@@ -682,6 +686,83 @@ TEST(test_amf_nas5g_pkt_process, test_amf_security_mode_reject_message_data) {
+  bstring buffer;
+  uint32_t bytes = 0;
+
+  int len = construct_pdu_session_reject_dl_req(
```
Is it possible to call "amf_pdu_session_establishment_reject". amf_send_msg_to_task() gets stubbed. So we can make use of it.

sreedharkumartn replied:
Zenhub ticket #9944 is handling this comment.

sreedharkumartn committed:
Addressed review comments

Signed-off-by: sreedharkumartn <sreedhar.kumar@wavelabs.ai>

sreedharkumartn commented:
- Verified below Failures cases Using UERANSIM.
- Attached Pcap snap and log file of all failure cases.
1. PDU Session Establishment Failure with reason 27
Log file:
[mme_missing_dnn.log](https://github.com/magma/magma/files/7462326/mme_missing_dnn.log)
Pcap Snap:
![Missing_dnn](https://user-images.githubusercontent.com/89978170/139902945-3647fdc6-7e65-49b4-a9cc-2ac520d7fb82.JPG)
2. PDU Session Establishment Failure with reason 28
Log file:
[mme_unknown_pdu_type.log](https://github.com/magma/magma/files/7462339/mme_unknown_pdu_type.log)
Pcap Snap:
![Unknown_PDU_type](https://user-images.githubusercontent.com/89978170/139903725-7587a62a-d5d0-45ed-8c5d-059100250d84.JPG)
3. PDU Session Establishment Failure with reason 43
Log file:
[mme_invalid_pdu_identity.log](https://github.com/magma/magma/files/7462351/mme_invalid_pdu_identity.log)
Pcap snap:
![Invalid_PDU_Session_Identity](https://user-images.githubusercontent.com/89978170/139904219-ed667c3b-8be4-4b51-ba19-7d2ae09e3065.JPG)
UT cases validated:
![UT_Test](https://user-images.githubusercontent.com/89978170/139904327-19ea0400-a275-4343-8a08-837b5f4313b8.JPG)

panyogesh commented:
Comments seems to be addressed.

ulaskozat commented:
@ssanadhya please steer this PR.

ssanadhya commented on lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp:
```diff
@@ -563,7 +562,100 @@ int amf_smf_send(
+      msg->payload_container.smf_msg.msg.pdu_session_estab_request
+          .pdu_session_type.type_val);
+  if (M5GPduSessionType::UNSTRUCTURED == session_type ||
+      M5GPduSessionType::ETHERNET == session_type) {
```
Please use the variable on the left hand side of comparison for readability
```suggestion
  if (session_type == M5GPduSessionType::UNSTRUCTURED  ||
      session_type ==  M5GPduSessionType::ETHERNET) {
```

sreedharkumartn replied:
Addressed review comment.

ssanadhya commented on lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp:
```diff
@@ -563,7 +562,100 @@ int amf_smf_send(
+
+  if ((msg->payload_container.smf_msg.header.message_type ==
+       PDU_SESSION_ESTABLISHMENT_REQUEST) &&
+      (M5GRequestType::INITIAL_REQUEST == requestType) && smf_ctx) {
```
```suggestion
      (requestType == M5GRequestType::INITIAL_REQUEST) && smf_ctx) {
```

sreedharkumartn replied:
Addressed review comment.

ssanadhya commented on lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp:
```diff
@@ -563,7 +562,100 @@ int amf_smf_send(
+      smf_ctx->duplicate_pdu_session_est_req_count += 1;
+      OAILOG_INFO(
+          LOG_AMF_APP,
+          "duplicate Initial PDU Session establishment request received");
```
```suggestion
          "Duplicate Initial PDU Session establishment request received");
```

sreedharkumartn replied:
Addressed review comment

ssanadhya commented on lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp:
```diff
@@ -563,7 +562,100 @@ int amf_smf_send(
+**                                                                        **
+**                                                                        **
+***************************************************************************/
+M5GSmCause amf_smf_validate_ie_params(
```
This function is not validating all ies in message, just checking for session establishment failure cause. Can we instead call it `amf_smf_get_smcause()`

sreedharkumartn replied:
Addressed review comment.

ssanadhya commented on lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp:
```diff
@@ -563,7 +562,100 @@ int amf_smf_send(
+**                                                                        **
+** Name:    amf_smf_validate_context()                                    **
+**                                                                        **
+** Description: function to handle 5GMM cause                             **
```
```suggestion
** Description: function to check if max PDU sessions reached             **
```

sreedharkumartn replied:
Addressed review comment

ssanadhya commented on lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp:
```diff
@@ -362,13 +359,13 @@ static int pdu_session_resource_release_t3592_handler(
-int amf_smf_send(
+int amf_smf_process_pdu_session_packet(
     amf_ue_ngap_id_t ue_id, ULNASTransportMsg* msg, int amf_cause) {
   int rc = 1;
```
Please define a macro to indicate what 1 means (success/failure).

sreedharkumartn replied:
Addressed review comment.

ssanadhya commented on lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp:
```diff
@@ -889,4 +981,182 @@ int handle_sm_message_routing_failure(
+  pdu_sess_est_reject.header.pdu_session_id = session_id;
+  pdu_sess_est_reject.header.message_type   = PDU_SESSION_ESTABLISHMENT_REJECT;
+  pdu_sess_est_reject.header.procedure_transaction_id = pti;
+  // container_len++;
```
Remove commented code

sreedharkumartn replied:
Addressed review comment.

ssanadhya commented on lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp:
```diff
@@ -889,4 +981,182 @@ int handle_sm_message_routing_failure(
+
+  if (bytes > 0) {
+    buffer->slen = bytes;
+    amf_app_handle_nas_dl_req(ue_id, buffer, rc);
```
Is the memory allocated for `buffer` always released inside `amf_app_handle_nas_dl_req`?

sreedharkumartn replied:
Yes, Memory allocated for buffer always released inside amf_app_handle_nas_dl_req.
![buffer destroy](https://user-images.githubusercontent.com/89978170/141515964-b578471b-55f9-472f-ac95-4ea500391c18.JPG)

ssanadhya replied:
@sreedharkumartn , note that the bstring nas_msg is not destroyed if the function returns due to null value of amf_app_desc_p or null ue_context. Can you create a separate PR to destroy nas_msg in those cases as well?

ssanadhya reviewed (changes requested):
@sreedharkumartn , can you also add unit tests for other functions added in this pull request, such as `amf_smf_validate_ie_params` ?

ulaskozat commented on lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp:
```diff
@@ -889,4 +981,182 @@ int handle_sm_message_routing_failure(
+ **  Return     :  len                 : buffer required for               **
+ **                                      DLNASTransportMsg                 **
+ **                                                                        **
+ ***************************************************************************/
```
fix these lines

sreedharkumartn replied:
Addressed review comment.

ulaskozat commented on lte/gateway/c/core/oai/tasks/nas5g/src/SmfMessage.cpp:
```diff
@@ -103,6 +103,11 @@ int SmfMsg::SmfMsgDecodeMsg(SmfMsg* msg, uint8_t* buffer, uint32_t len) {
       break;
+    case PDU_SESSION_ESTABLISHMENT_REJECT:
+      decode_result = msg->msg.pdu_session_estab_reject
+                          .DecodePDUSessionEstablishmentRejectMsg(
```
is this covered by any test case?

sreedharkumartn replied:
![pdu_session_establish_reject](https://user-images.githubusercontent.com/89978170/141928147-4a2d814b-9de7-43b2-b98a-c38a15c2e3da.JPG)
Yes, this is covered in the above test case.

ulaskozat commented on lte/gateway/c/core/oai/tasks/nas5g/src/M5GPDUSessionEstablishmentReject.cpp:
```diff
@@ -21,8 +21,51 @@ PDUSessionEstablishmentRejectMsg::~PDUSessionEstablishmentRejectMsg(){};
 int PDUSessionEstablishmentRejectMsg::DecodePDUSessionEstablishmentRejectMsg(
```
is this covered by a test case?

sreedharkumartn replied:
![pdu_session_establish_reject](https://user-images.githubusercontent.com/89978170/141928277-e08c77e6-19de-411c-8ccb-beb3e2264a09.JPG)
This is covered in the above test case.

ssanadhya commented:
> * Verified below Failures cases Using UERANSIM.
> * Attached Pcap snap and log file of all failure cases.
> 
> 1. PDU Session Establishment Failure with reason 27
>    Log file:
>    [mme_missing_dnn.log](https://github.com/magma/magma/files/7462326/mme_missing_dnn.log)
>    Pcap Snap:
>    ![Missing_dnn](https://user-images.githubusercontent.com/89978170/139902945-3647fdc6-7e65-49b4-a9cc-2ac520d7fb82.JPG)
> 2. PDU Session Establishment Failure with reason 28
>    Log file:
>    [mme_unknown_pdu_type.log](https://github.com/magma/magma/files/7462339/mme_unknown_pdu_type.log)
>    Pcap Snap:
>    ![Unknown_PDU_type](https://user-images.githubusercontent.com/89978170/139903725-7587a62a-d5d0-45ed-8c5d-059100250d84.JPG)
> 3. PDU Session Establishment Failure with reason 43
>    Log file:
>    [mme_invalid_pdu_identity.log](https://github.com/magma/magma/files/7462351/mme_invalid_pdu_identity.log)
>    Pcap snap:
>    ![Invalid_PDU_Session_Identity](https://user-images.githubusercontent.com/89978170/139904219-ed667c3b-8be4-4b51-ba19-7d2ae09e3065.JPG)
>    UT cases validated:
>    ![UT_Test](https://user-images.githubusercontent.com/89978170/139904327-19ea0400-a275-4343-8a08-837b5f4313b8.JPG)

@sreedharkumartn , The pcap and MME log for cause 28 has no packets/logs related to PDU session establishment.

sreedharkumartn committed:
Addressed Review Comments (R2)

Signed-off-by: sreedharkumartn <sreedhar.kumar@wavelabs.ai>

sreedharkumartn commented:
@ssanadhya , added the test cases for three PDU Session Establishment rejection scenarios in amf_smf_get_smcause() function. Updated the PCAP and MME log files for three PDU Session Establishment rejection scenarios.
@ssanadhya and @ulaskozat , addressed the review comments.

- Verified the below failure cases with UERANSIM.
- Attached PCAP snap and log file of all failure cases.

1. PDU Session Establishment Failure with reason 27
    Case 1: UE DNN: None, Subscriberdb DNN: None.
    Log file:  
[mme_missing_dnn_ue_subscriberdb.log](https://github.com/magma/magma/files/7529422/mme_missing_dnn_ue_subscriberdb.log)
    PCAP snap:
![missing_dnn_ue_subscriberdb](https://user-images.githubusercontent.com/89978170/141510872-30840a4c-fc3e-4b32-a90b-066bedfbdc11.JPG)
    Case 2: UE have DNN, Subscriberdb DNN: None.
    Log file:
[mme_with_dnn_ue_missing_dnn_subscriberdb.log](https://github.com/magma/magma/files/7529555/mme_with_dnn_ue_missing_dnn_subscriberdb.log)
    PCAP snap:
![with_dnn_ue_missing_dnn_subscriberdb](https://user-images.githubusercontent.com/89978170/141512857-058aab2c-bf9b-40b9-89f3-b99580ea0c71.JPG)
    Case 3: UE DNN: None, Subscriberdb have DNN.
     Log file:
[mme_missing_dnn_ue_subscriberdb.log](https://github.com/magma/magma/files/7529560/mme_missing_dnn_ue_subscriberdb.log)
    PCAP snap:
![missing_dnn_ue_with_dnn_subscriberdb](https://user-images.githubusercontent.com/89978170/141514366-1e2cf394-641f-4615-aa9c-c61d33e0459b.JPG)
2. PDU Session Establishment Failure with reason 28
Log file:
[mme_unknown_pdu_session_type.log](https://github.com/magma/magma/files/7529571/mme_unknown_pdu_session_type.log)
PCAP snap:
![unknown_pdu_session_type](https://user-images.githubusercontent.com/89978170/141514823-a0836ec9-f516-4dd0-a729-fa02a4bb00e8.JPG)
3. PDU Session Establishment Failure with reason 43
Log file:
[mme_invalid_pdu_session_identity.log](https://github.com/magma/magma/files/7529586/mme_invalid_pdu_session_identity.log)
PCAP snap:
![invalid_pdu_session_identity](https://user-images.githubusercontent.com/89978170/141514928-4d0dc4f2-a4d8-43ed-af89-8c3123e61080.JPG)

UT cases validated with added test cases.
![UT_Test_cases](https://user-images.githubusercontent.com/89978170/141515079-c9185f8c-ecc9-4539-8743-c26e90f40efa.JPG)

ulaskozat commented:
@ssanadhya could you check if your review comments are all addressed?

ssanadhya commented on lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp:
```diff
@@ -563,7 +562,104 @@ int amf_smf_send(
+  */
+  if (msg->dnn.dnn.empty() &&
+      (ue_context->amf_context.apn_config_profile.nb_apns == 0)) {
+    cause = M5GSmCause::MISSING_OR_UNKNOWN_DNN;
```
You can return from here, no?

sreedharkumartn replied:
Addressed review comment

ssanadhya commented on lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp:
```diff
@@ -563,7 +562,104 @@ int amf_smf_send(
+          .pdu_session_type.type_val);
+  if (session_type == M5GPduSessionType::UNSTRUCTURED ||
+      session_type == M5GPduSessionType::ETHERNET) {
+    cause = M5GSmCause::UNKNOWN_PDU_SESSION_TYPE;
```
Return from here

sreedharkumartn replied:
Addressed review comment.

ssanadhya commented on lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp:
```diff
@@ -563,7 +562,104 @@ int amf_smf_send(
+       (M5GRequestType::EXISTING_PDU_SESSION == requestType))) {
+    if (ue_context->amf_context.smf_ctxt_map.size() >=
+        MAX_UE_PDU_SESSION_LIMIT) {
+      M5GMmCause mm_cause = M5GMmCause::MAX_PDU_SESSIONS_REACHED;
```
Duplicate declaration of mm_cause, already done in line 640 above

sreedharkumartn replied:
Addressed review comment.

ssanadhya commented on lte/gateway/c/core/oai/tasks/amf/include/amf_smf_packet_handler.h:
```diff
@@ -0,0 +1,36 @@
+#include "lte/gateway/c/core/oai/tasks/nas5g/include/M5gNasMessage.h"
+
+namespace magma5g {
+#define PAYLOAD_CONTAINER_LENGTH 2
```
```suggestion
#define PAYLOAD_CONTAINER_TAG_LENGTH 2
```

sreedharkumartn replied:
Addressed review comment

ssanadhya commented on lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp:
```diff
@@ -889,4 +985,181 @@ int handle_sm_message_routing_failure(
+  container_len++;
+
+  dlmsg->payload_container.len = container_len;
+  len += PAYLOAD_CONTAINER_LENGTH;
```
```suggestion
  len += PAYLOAD_CONTAINER_TAG_LENGTH;
```

sreedharkumartn replied:
Addressed review comment.

ssanadhya commented on lte/gateway/c/core/oai/tasks/amf/include/amf_smf_packet_handler.h:
```diff
@@ -0,0 +1,36 @@
+#define PAYLOAD_CONTAINER_LENGTH 2
+#define AMF_CAUSE_SUCCESS 1
+const uint32_t MAX_UE_PDU_SESSION_LIMIT                             = 15;
+const uint32_t MAX_UE_INITIAL_PDU_SESSION_ESTABLISHMENT_REQ_ALLOWED = 5;
```
For consistency, please use #define instead of `const uint32_t`

sreedharkumartn replied:
Addressed review comment.

ssanadhya commented on lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp:
```diff
@@ -842,6 +879,115 @@ TEST(test_dnn, test_amf_validate_dnn) {
+  bool decode_res = false;
+  memset(&pdu_sess_est_req, 0, sizeof(ULNASTransportMsg));
+
+  decode_res = decode_ul_nas_transport_msg(&pdu_sess_est_req, pdu, len);
```
Please add a check that decode was successful

sreedharkumartn replied:
Addressed review comment.

ssanadhya commented on lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp:
```diff
@@ -842,6 +879,115 @@ TEST(test_dnn, test_amf_validate_dnn) {
+
+  decode_res = decode_ul_nas_transport_msg(&pdu_sess_est_req, pdu, len);
+
+  // build uplinknastransport
```
Irrelevant comment

sreedharkumartn replied:
Addressed review comment.

ssanadhya commented on lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp:
```diff
@@ -842,6 +879,115 @@ TEST(test_dnn, test_amf_validate_dnn) {
+
+  decode_res = decode_ul_nas_transport_msg(&pdu_sess_est_req, pdu, len);
+
+  // build uplinknastransport
```
Irrelevant comment

sreedharkumartn replied:
Addressed review comment.

ssanadhya commented on lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp:
```diff
@@ -842,6 +879,115 @@ TEST(test_dnn, test_amf_validate_dnn) {
+  bool decode_res = false;
+  memset(&pdu_sess_est_req, 0, sizeof(ULNASTransportMsg));
+
+  decode_res = decode_ul_nas_transport_msg(&pdu_sess_est_req, pdu, len);
```
Add a check that decode was successful

sreedharkumartn replied:
Addressed review comment.

ssanadhya commented on lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp:
```diff
@@ -842,6 +879,115 @@ TEST(test_dnn, test_amf_validate_dnn) {
+  decode_res = decode_ul_nas_transport_msg(&pdu_sess_est_req, pdu, len);
+
+  EXPECT_EQ(decode_res, true);
+  // build uplinknastransport
```
Irrelevant comment

sreedharkumartn replied:
Addressed review comment.

ssanadhya commented on lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp:
```diff
@@ -842,6 +879,115 @@ TEST(test_dnn, test_amf_validate_dnn) {
+  M5GSmCause cause_2   = amf_smf_get_smcause(ue_id, &pdu_sess_est_req);
+  M5GSmCause cause_3   = amf_smf_get_smcause(ue_id, &pdu_sess_est_req);
+  M5GSmCause cause_4   = amf_smf_get_smcause(ue_id, &pdu_sess_est_req);
+  M5GSmCause cause_5   = amf_smf_get_smcause(ue_id, &pdu_sess_est_req);
```
```suggestion
  for (int req_cnt=0; req_cnt < MAX_UE_INITIAL_PDU_SESSION_ESTABLISHMENT_REQ_ALLOWED; req_cnt++) {
  amf_smf_get_smcause(ue_id, &pdu_sess_est_req);
  }
```

sreedharkumartn replied:
Addressed review comment.

ssanadhya commented on lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp:
```diff
@@ -842,6 +879,115 @@ TEST(test_dnn, test_amf_validate_dnn) {
+  M5GSmCause cause_4   = amf_smf_get_smcause(ue_id, &pdu_sess_est_req);
+  M5GSmCause cause_5   = amf_smf_get_smcause(ue_id, &pdu_sess_est_req);
+  M5GSmCause cause_dup = amf_smf_get_smcause(ue_id, &pdu_sess_est_req);
+  EXPECT_EQ(M5GSmCause::INVALID_PDU_SESSION_IDENTITY, cause_dup);
```
```suggestion
  EXPECT_EQ(cause_dup, M5GSmCause::INVALID_PDU_SESSION_IDENTITY);
```

sreedharkumartn replied:
Addressed review comment.

ssanadhya reviewed (approved):
LGTM, with minor comments. Please address these before merging.

sreedharkumartn committed:
Addressed Review Comments (R2)

Signed-off-by: sreedharkumartn <sreedhar.kumar@wavelabs.ai>

sreedharkumartn commented:
@ssanadhya and @ulaskozat , addressed the review comments.

- Verified the below failure cases with UERANSIM.
- Attached PCAP snap and log file of all failure cases.

1. PDU Session Establishment Failure with reason 27
    Case 1: UE DNN: None, Subscriberdb DNN: None.
    Log file:    
[mme_missing_dnn_case_1.log](https://github.com/magma/magma/files/7551311/mme_missing_dnn_case_1.log)
    PCAP snap:
![missing_dnn_case_1](https://user-images.githubusercontent.com/89978170/142121836-61e28955-75c7-4d1b-8818-b8bf70cf8736.JPG)
    Case 2: UE have DNN, Subscriberdb DNN: None.
    Log file:
[mme_missing_dnn_case_2.log](https://github.com/magma/magma/files/7551341/mme_missing_dnn_case_2.log)
    PCAP snap:
![missing_dnn_case_2](https://user-images.githubusercontent.com/89978170/142122503-7b87743d-761b-4537-a307-f3e5d8cb2f5e.JPG)
    Case 3: UE DNN: None, Subscriberdb have DNN.
     Log file:
[mme_missing_dnn_case_3.log](https://github.com/magma/magma/files/7551342/mme_missing_dnn_case_3.log)
     PCAP snap:
![missing_dnn_case_3](https://user-images.githubusercontent.com/89978170/142123614-6d7805fc-a784-4521-886f-1ba0926856bc.JPG)
2. PDU Session Establishment Failure with reason 28
Log file:
[mme_unknown_pdu_session_type.log](https://github.com/magma/magma/files/7551343/mme_unknown_pdu_session_type.log)
PCAP snap:
![unknown_pdu_session_type](https://user-images.githubusercontent.com/89978170/142124250-09a2a006-6752-4f19-b22a-0e55a7406cd2.JPG)
3. PDU Session Establishment Failure with reason 43
Log file:
[mme_invalid_pdu_session_identity.log](https://github.com/magma/magma/files/7551345/mme_invalid_pdu_session_identity.log)
PCAP snap:
![invalid_pdu_session_identity](https://user-images.githubusercontent.com/89978170/142124668-8d11483a-70d4-4240-ac06-938682a9d628.JPG)

UT cases validated.
![UT_cases](https://user-images.githubusercontent.com/89978170/142125055-6be6f632-551c-4d5d-8185-38528f0bb007.JPG)
